cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java

```diff
@@ -91,51 +201,730 @@ public void schedule(Id.Program program, SchedulableProgramType programType, Ite
-    return String.format("%s:%s:%s:%s", program.getNamespaceId(), program.getApplicationId(),
-                         programType.name(), program.getId());
+  /**
+   * One instance of this class contains a list of {@link StreamSizeSchedule}s, which are all interested
+   * in the same stream. This instance subscribes to the size notification of the stream, and polls the
+   * stream for its size whenever the schedules it references need the information.
+   * The {@link StreamSizeScheduler} communicates with this class, which in turn communicates to the schedules
+   * it contains to perform operations on the schedules - suspend, resume, etc.
+   */
+  private final class StreamSubscriber implements NotificationHandler<StreamSizeNotification>, Cancellable {
+    // Key is the schedule ID
+    private final ConcurrentMap<String, StreamSizeScheduleTask> scheduleTasks;
+    private final Id.Stream streamId;
+    private final AtomicInteger activeTasks;
+
+    private Cancellable notificationSubscription;
+    private ScheduledFuture<?> scheduledPolling;
+    private StreamSizeNotification lastNotification;
+    private StreamSizeNotification lastPollingInfo;
+
+    private StreamSubscriber(Id.Stream streamId) {
+      this.streamId = streamId;
+      this.scheduleTasks = Maps.newConcurrentMap();
+      this.activeTasks = new AtomicInteger(0);
+    }
+
+    public void start() throws NotificationFeedException, NotificationFeedNotFoundException {
+      notificationSubscription = notificationService.subscribe(getFeed(), this, sendPollingInfoExecutor);
+    }
+
+    @Override
+    public void cancel() {
+      if (scheduledPolling != null) {
+        scheduledPolling.cancel(true);
+      }
+      if (notificationSubscription != null) {
+        notificationSubscription.cancel();
+      }
+    }
+
+    /**
+     * Add a new scheduling task based on the data received by the stream referenced by {@code this} object.
+     * @return {@code true} if the task was created successfully, {@code false} if it already exists
+     */
+    public boolean createScheduleTask(Id.Program programId, SchedulableProgramType programType,
+                                      StreamSizeSchedule streamSizeSchedule, boolean active,
+                                      long baseRunSize, long baseRunTs, boolean persist) throws SchedulerException {
+      final StreamSizeScheduleTask newTask = new StreamSizeScheduleTask(programId, programType, streamSizeSchedule);
+      synchronized (this) {
+        StreamSizeScheduleTask previous =
+          scheduleTasks.putIfAbsent(AbstractSchedulerService.scheduleIdFor(programId, programType,
+                                                                           streamSizeSchedule.getName()),
+                                    newTask);
+        if (previous != null) {
+          // We cannot replace an existing schedule - that functionality is not wanted - yet
+          return false;
```

Comment:
this should throw an exception. Or the caller should. If I understand correctly, we support updating a schedule, but through a different means.